File: workspaces/decision-platform/research/s1-4x-numeric-parity/haskell/src/core/S14X/Core/ProductionMetrics.hs
module S14X.Core.ProductionMetrics
  ( annualizedVolatility,
    cagr,
    cumulativeReturn,
    historicalCvar,
    historicalVar,
    logReturns,
    maxDrawdown,
    realizedVolatility,
    sharpeRatio,
    simpleReturns,
    sortinoRatio,
  )
where

import           Numeric.SpecFunctions (expm1)

import qualified Data.Vector.Unboxed as U

import           S14X.Core.Error (StableError (ConfidenceInvalid, DenominatorZero, EquityInitialNonPositive, EquityNegative, PeriodsPerYearInvalid, PricesNonPositive, ResultNonFinite, RiskFreeRateInvalid, SimpleReturnBelowMinusOne, TailEmpty, TargetReturnInvalid))
import           S14X.Core.NumericPrimitives (hf7Quantile, meanVector, sampleVariance, sumVector)
import           S14X.Core.ScalarValidation (ensureFinite, validateConfidence, validateFiniteScalar,
                                             validatePositiveInteger)
import           S14X.Core.Validation (validateProductionVector)

-- | 양수 가격 벡터를 인접 단순수익률 벡터로 변환한다.
-- 길이·유한성·양수 계약 위반과 비유한 결과는 stable production 오류로 반환한다.
simpleReturns :: U.Vector Double -> Either StableError (U.Vector Double)
simpleReturns rawPrices = do
  prices <- validateProductionVector 2 rawPrices
  if U.any (<= 0.0) prices
    then Left PricesNonPositive
    else ensureFiniteVector (U.zipWith (\previous current -> current / previous - 1.0) prices (U.drop 1 prices))

-- | 양수 가격 벡터의 인접 로그수익률을 계산한다.
-- 입력을 복사하거나 변경하지 않으며 가격·결과 검증 실패를 'StableError'로 닫는다.
logReturns :: U.Vector Double -> Either StableError (U.Vector Double)
logReturns rawPrices = do
  prices <- validateProductionVector 2 rawPrices
  if U.any (<= 0.0) prices
    then Left PricesNonPositive
    else
      ensureFiniteVector
        (U.zipWith (\previous current -> log current - log previous) prices (U.drop 1 prices))

-- | 단순수익률 경로의 복리 누적수익률을 반환한다.
-- @-1@은 전액 손실로 허용하지만 그보다 작은 수익률과 비유한 결과는 거부한다.
cumulativeReturn :: U.Vector Double -> Either StableError Double
cumulativeReturn rawReturns = do
  returns <- validateProductionVector 1 rawReturns
  if U.any (< -1.0) returns
    then Left SimpleReturnBelowMinusOne
    else
      if U.any (== -1.0) returns
        then Right (-1.0)
        else ensureFinite ResultNonFinite (U.foldl' (\total value -> total * (1.0 + value)) 1.0 returns - 1.0)

-- | 양수 가격 경로와 연간 관측 수로 연복리성장률을 계산한다.
-- arbitrary-size 'Integer' 주기는 양수여야 하며 최종 Float64 overflow는 결과 오류가 된다.
cagr :: U.Vector Double -> Integer -> Either StableError Double
cagr rawPrices periodsPerYear = do
  prices <- validateProductionVector 2 rawPrices
  periods <- validatePositiveInteger PeriodsPerYearInvalid periodsPerYear
  if U.any (<= 0.0) prices
    then Left PricesNonPositive
    else
      case (prices U.!? 0, prices U.!? (U.length prices - 1)) of
        (Just initialPrice, Just finalPrice) ->
          let annualization = periods / fromIntegral (U.length prices - 1)
              logGrowth = log finalPrice - log initialPrice
           in ensureFinite ResultNonFinite (expm1 (annualization * logGrowth))
        _ -> Left ResultNonFinite

-- | 최소 두 수익률의 표본표준편차를 실현 변동성으로 반환한다.
-- production 벡터 검증과 결과 유한성 계약을 먼저 적용한다.
realizedVolatility :: U.Vector Double -> Either StableError Double
realizedVolatility rawReturns = do
  returns <- validateProductionVector 2 rawReturns
  ensureFinite ResultNonFinite (sqrt (sampleVariance returns))

-- | 실현 변동성을 양의 연간 관측 수의 제곱근으로 연율화한다.
-- 입력·주기·Float64 결과 오류는 정해진 production 오류 우선순위로 반환한다.
annualizedVolatility :: U.Vector Double -> Integer -> Either StableError Double
annualizedVolatility rawReturns periodsPerYear = do
  returns <- validateProductionVector 2 rawReturns
  periods <- validatePositiveInteger PeriodsPerYearInvalid periodsPerYear
  ensureFinite ResultNonFinite (sqrt (sampleVariance returns) * sqrt periods)

-- | 비음수 자산가치 경로에서 running peak 대비 최저 drawdown을 계산한다.
-- 최초 값은 양수여야 하고 이후 음수 값과 비유한 입력은 stable 오류로 거부한다.
maxDrawdown :: U.Vector Double -> Either StableError Double
maxDrawdown rawEquity = do
  equity <- validateProductionVector 1 rawEquity
  case equity U.!? 0 of
    Nothing -> Left ResultNonFinite
    Just initial
      | initial <= 0.0 -> Left EquityInitialNonPositive
      | U.any (< 0.0) (U.drop 1 equity) -> Left EquityNegative
      | otherwise ->
          let (_, drawdown) =
                U.foldl'
                  ( \(peak, minimumDrawdown) value ->
                      let nextPeak = max peak value
                          currentDrawdown = value / nextPeak - 1.0
                       in (nextPeak, min minimumDrawdown currentDrawdown)
                  )
                  (initial, 0.0)
                  equity
           in ensureFinite ResultNonFinite drawdown

-- | 수익률, 무위험수익률, 양의 연간 주기로 연율화 Sharpe ratio를 계산한다.
-- 표본분산이 0이면 나눗셈 대신 'DenominatorZero'를 반환한다.
sharpeRatio ::
  U.Vector Double ->
  Double ->
  Integer ->
  Either StableError Double
sharpeRatio rawReturns riskFreeRate periodsPerYear = do
  returns <- validateProductionVector 2 rawReturns
  riskFree <- validateFiniteScalar RiskFreeRateInvalid riskFreeRate
  periods <- validatePositiveInteger PeriodsPerYearInvalid periodsPerYear
  let excess = U.map (\value -> value - riskFree) returns
      denominator = sqrt (sampleVariance excess)
  if denominator == 0.0
    then Left DenominatorZero
    else ensureFinite ResultNonFinite (meanVector excess / denominator * sqrt periods)

-- | 수익률, 목표수익률, 양의 연간 주기로 Sortino ratio를 계산한다.
-- downside RMS가 0이면 결과를 만들지 않고 'DenominatorZero'로 닫는다.
sortinoRatio ::
  U.Vector Double ->
  Double ->
  Integer ->
  Either StableError Double
sortinoRatio rawReturns targetReturn periodsPerYear = do
  returns <- validateProductionVector 2 rawReturns
  target <- validateFiniteScalar TargetReturnInvalid targetReturn
  periods <- validatePositiveInteger PeriodsPerYearInvalid periodsPerYear
  let excess = U.map (\value -> value - target) returns
      downside = U.map (`min` 0.0) excess
      denominator = sqrt (sumVector (U.map (\value -> value * value) downside) / fromIntegral (U.length downside))
  if denominator == 0.0
    then Left DenominatorZero
    else ensureFinite ResultNonFinite (meanVector excess / denominator * sqrt periods)

-- | production 수익률의 하위 꼬리를 HF7 quantile로 평가해 historical VaR를 반환한다.
-- confidence는 열린 구간 @(0,1)@이어야 한다.
historicalVar :: U.Vector Double -> Double -> Either StableError Double
historicalVar rawReturns confidence = do
  returns <- validateProductionVector 2 rawReturns
  probability <- validateConfidence ConfidenceInvalid confidence
  hf7Quantile returns (1.0 - probability)

-- | HF7 VaR 이하 관측값의 보상합 평균으로 historical CVaR를 계산한다.
-- 비어 있는 꼬리와 비유한 결과는 각각 명시된 stable 오류가 된다.
historicalCvar :: U.Vector Double -> Double -> Either StableError Double
historicalCvar rawReturns confidence = do
  returns <- validateProductionVector 2 rawReturns
  probability <- validateConfidence ConfidenceInvalid confidence
  threshold <- hf7Quantile returns (1.0 - probability)
  let tailValues = U.filter (<= threshold) returns
  if U.null tailValues
    then Left TailEmpty
    else ensureFinite ResultNonFinite (meanVector tailValues)

ensureFiniteVector :: U.Vector Double -> Either StableError (U.Vector Double)
ensureFiniteVector values
  | U.any (\value -> isNaN value || isInfinite value) values = Left ResultNonFinite
  | otherwise = Right values

File: workspaces/decision-platform/research/s1-4x-numeric-parity/scala/src/main/scala/ai/trading/coach/s14x/core/ProductionMetrics.scala
package ai.trading.coach.s14x.core

object ProductionMetrics:
  /** JSON semantic precedence용 raw boundary이며 성공 시 defensive immutable snapshot을 쓴다. */
  def simpleReturnsRaw(raw: Any): Either[StableError, Vector[Double]] =
    Validation.productionRawSequence(raw, minimumLength = 2).flatMap(simpleReturns)

  /** 양수 가격 시계열을 받아 인접 단순수익률을 반환하며, frozen production 오류 우선순위를 보존한다. */
  def simpleReturns(prices: Vector[Double]): Either[StableError, Vector[Double]] =
    Validation.productionSequence(prices, minimumLength = 2).flatMap { values =>
      if values.exists(_ <= 0.0) then Left(StableError.PricesNonPositive)
      else
        val result = values.zip(values.drop(1)).map { case (previous, current) =>
          NumericPrimitives.normalizeZero(current / previous - 1.0)
        }
        if result.forall(_.isFinite) then Right(result) else Left(StableError.ResultNonFinite)
    }

  /** 양수 immutable 가격 시계열을 검증한 뒤 인접 로그수익률을 반환한다. */
  def logReturns(prices: Vector[Double]): Either[StableError, Vector[Double]] =
    Validation.productionSequence(prices, minimumLength = 2).flatMap { values =>
      if values.exists(_ <= 0.0) then Left(StableError.PricesNonPositive)
      else
        val result = values.zip(values.drop(1)).map { case (previous, current) =>
          NumericPrimitives.normalizeZero(math.log(current) - math.log(previous))
        }
        if result.forall(_.isFinite) then Right(result) else Left(StableError.ResultNonFinite)
    }

  /** 단순수익률 시계열을 복리 누적하고, -1 미만 입력과 비유한 결과를 stable error로 반환한다. */
  def cumulativeReturn(returns: Vector[Double]): Either[StableError, Double] =
    Validation.productionSequence(returns, minimumLength = 1).flatMap { values =>
      if values.exists(_ < -1.0) then Left(StableError.SimpleReturnBelowMinusOne)
      else if values.exists(_ == -1.0) then Right(-1.0)
      else
        val result = values.foldLeft(1.0)((product, value) => product * (1.0 + value)) - 1.0
        Validation.finiteProduction(result)
    }

  /** 양수 가격 경로와 양의 연환산 주기를 받아 기하 연환산 수익률을 계산한다. */
  def cagr(
      prices: Vector[Double],
      periodsPerYear: BigInt = BigInt(252)
  ): Either[StableError, Double] =
    for
      values <- Validation.productionSequence(prices, minimumLength = 2)
      periods <- Validation.positivePeriods(periodsPerYear)
      result <-
        if values.exists(_ <= 0.0) then Left(StableError.PricesNonPositive)
        else
          val first = values.take(1).foldLeft(Double.NaN)((_, value) => value)
          val finalValue = values.drop(values.size - 1).foldLeft(Double.NaN)((_, value) => value)
          val annualization = periods.toDouble / (values.size - 1).toDouble
          Validation.finiteProduction(
            math.expm1(annualization * (math.log(finalValue) - math.log(first)))
          )
    yield result

  /** 두 개 이상의 로그수익률을 표본 표준편차로 집계해 실현 변동성을 반환한다. */
  def realizedVolatility(logReturns: Vector[Double]): Either[StableError, Double] =
    Validation
      .productionSequence(logReturns, minimumLength = 2)
      .flatMap(values =>
        Validation.finiteProduction(NumericPrimitives.sampleStandardDeviation(values))
      )

  /** 로그수익률 표본 변동성을 양의 연환산 주기의 제곱근으로 확장한다. */
  def annualizedVolatility(
      logReturns: Vector[Double],
      periodsPerYear: BigInt = BigInt(252)
  ): Either[StableError, Double] =
    for
      values <- Validation.productionSequence(logReturns, minimumLength = 2)
      periods <- Validation.positivePeriods(periodsPerYear)
      result <- Validation.finiteProduction(
        NumericPrimitives.sampleStandardDeviation(values) * math.sqrt(periods.toDouble)
      )
    yield result

  /** 비음수 자산곡선의 running peak 대비 최소 수익률을 최대낙폭으로 반환한다. */
  def maxDrawdown(equityCurve: Vector[Double]): Either[StableError, Double] =
    Validation.productionSequence(equityCurve, minimumLength = 1).flatMap { values =>
      val first = values.take(1).foldLeft(Double.NaN)((_, value) => value)
      if first <= 0.0 then Left(StableError.EquityInitialNonPositive)
      else if values.drop(1).exists(_ < 0.0) then Left(StableError.EquityNegative)
      else
        val (_, minimumDrawdown) =
          values.foldLeft((first, 0.0)) { case ((runningPeak, minimum), value) =>
            val nextPeak = math.max(runningPeak, value)
            val drawdown = value / nextPeak - 1.0
            (nextPeak, math.min(minimum, drawdown))
          }
        Validation.finiteProduction(minimumDrawdown)
    }

  /** 수익률·주기당 무위험수익률·연환산 주기를 받아 표본 Sharpe 비율을 반환한다. */
  def sharpeRatio(
      returns: Vector[Double],
      riskFreeRate: Double = 0.0,
      periodsPerYear: BigInt = BigInt(252)
  ): Either[StableError, Double] =
    for
      values <- Validation.productionSequence(returns, minimumLength = 2)
      riskFree <-
        if riskFreeRate.isFinite then Right(riskFreeRate)
        else Left(StableError.RiskFreeRateInvalid)
      periods <- Validation.positivePeriods(periodsPerYear)
      excess = values.map(_ - riskFree)
      denominator = NumericPrimitives.sampleStandardDeviation(excess)
      result <-
        if denominator == 0.0 then Left(StableError.DenominatorZero)
        else
          Validation.finiteProduction(
            NumericPrimitives.compensatedSum(excess) / excess.size.toDouble / denominator *
              math.sqrt(periods.toDouble)
          )
    yield result

  /** 수익률·목표수익률·연환산 주기를 받아 전체 관측치 기준 downside Sortino 비율을 반환한다. */
  def sortinoRatio(
      returns: Vector[Double],
      targetReturn: Double = 0.0,
      periodsPerYear: BigInt = BigInt(252)
  ): Either[StableError, Double] =
    for
      values <- Validation.productionSequence(returns, minimumLength = 2)
      target <-
        if targetReturn.isFinite then Right(targetReturn)
        else Left(StableError.TargetReturnInvalid)
      periods <- Validation.positivePeriods(periodsPerYear)
      excess = values.map(_ - target)
      downside = excess.map(value => math.min(value, 0.0))
      denominator = math.sqrt(
        NumericPrimitives.compensatedSum(downside.map(value => value * value)) /
          downside.size.toDouble
      )
      result <-
        if denominator == 0.0 then Left(StableError.DenominatorZero)
        else
          Validation.finiteProduction(
            NumericPrimitives.compensatedSum(excess) / excess.size.toDouble / denominator *
              math.sqrt(periods.toDouble)
          )
    yield result

  private def historicalVarKernel(values: Vector[Double], confidence: Double): Double =
    val ordered = values.sorted
    val h = (ordered.size - 1).toDouble * (1.0 - confidence)
    val lowerIndex = math.floor(h).toInt
    val fraction = h - lowerIndex.toDouble
    val lower = ordered.lift(lowerIndex).fold(Double.NaN)(identity)
    val upper = ordered.lift(math.min(lowerIndex + 1, ordered.size - 1)).fold(Double.NaN)(identity)
    lower + fraction * (upper - lower)

  /** 수익률 표본과 신뢰수준을 받아 frozen 선형 보간 규칙의 historical VaR를 반환한다. */
  def historicalVar(
      returns: Vector[Double],
      confidence: Double = 0.95
  ): Either[StableError, Double] =
    for
      values <- Validation.productionSequence(returns, minimumLength = 2)
      probability <- Validation.productionConfidence(confidence)
      result <- Validation.finiteProduction(historicalVarKernel(values, probability))
    yield result

  /** 수익률 표본과 신뢰수준을 받아 VaR 이하 관측치 평균인 historical CVaR를 반환한다. */
  def historicalCvar(
      returns: Vector[Double],
      confidence: Double = 0.95
  ): Either[StableError, Double] =
    for
      values <- Validation.productionSequence(returns, minimumLength = 2)
      probability <- Validation.productionConfidence(confidence)
      threshold <- Validation.finiteProduction(historicalVarKernel(values, probability))
      tail = values.filter(_ <= threshold)
      result <-
        if tail.isEmpty then Left(StableError.TailEmpty)
        else
          Validation.finiteProduction(
            NumericPrimitives.compensatedSum(tail) / tail.size.toDouble
          )
    yield result
